File: astar.py
import heapq

class Node:
    def __init__(self, position, parent=None):
        self.position = position  # (x, y)
        self.parent = parent
        self.g = 0  # Cost from start to current node
        self.h = 0  # Heuristic cost to goal
        self.f = 0  # Total cost

    def __lt__(self, other):
        return self.f < other.f

def heuristic(a, b):
    # Manhattan distance
    return abs(a[0] - b[0]) + abs(a[1] - b[1])

def astar_search(grid, start, end):
    open_list = []
    closed_set = set()
    start_node = Node(start)
    end_node = Node(end)
    heapq.heappush(open_list, start_node)

    while open_list:
        current = heapq.heappop(open_list)

        if current.position == end_node.position:
            # Reconstruct path
            path = []
            while current:
                path.append(current.position)
                current = current.parent
            return path[::-1]  # Return reversed path

        closed_set.add(current.position)

        # Explore neighbors
        for dx, dy in [(-1,0), (1,0), (0,-1), (0,1)]:  # 4 directions
            neighbor_pos = (current.position[0] + dx, current.position[1] + dy)

            if (0 <= neighbor_pos[0] < len(grid) and
                0 <= neighbor_pos[1] < len(grid[0]) and
                grid[neighbor_pos[0]][neighbor_pos[1]] == 0 and
                neighbor_pos not in closed_set):

                neighbor = Node(neighbor_pos, current)
                neighbor.g = current.g + 1
                neighbor.h = heuristic(neighbor_pos, end)
                neighbor.f = neighbor.g + neighbor.h

                if any(n.position == neighbor.position and n.f <= neighbor.f for n in open_list):
                    continue

                heapq.heappush(open_list, neighbor)

    return None  # No path found

# Example grid (0 = walkable, 1 = wall)
grid = [
    [0, 0, 0, 0],
    [1, 1, 0, 1],
    [0, 0, 0, 0],
    [0, 1, 1, 0]
]

start = (0, 0)
end = (3, 3)
path = astar_search(grid, start, end)

print("Path:", path)

File: astar.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <cmath>
#include <algorithm>
#include <unordered_set>

using namespace std;

struct Node {
    int x, y;
    int g, h;
    Node* parent;

    Node(int x, int y, int g = 0, int h = 0, Node* p = nullptr)
        : x(x), y(y), g(g), h(h), parent(p) {}

    int f() const { return g + h; }

    bool operator>(const Node& other) const {
        return f() > other.f();
    }
};

int heuristic(int x1, int y1, int x2, int y2) {
    return abs(x1 - x2) + abs(y1 - y2); // Manhattan distance
}

bool isValid(int x, int y, const vector<vector<int>>& grid) {
    return x >= 0 && y >= 0 && x < grid.size() && y < grid[0].size()
           && grid[x][y] == 0;
}

vector<pair<int, int>> astar(const vector<vector<int>>& grid,
                              pair<int, int> start,
                              pair<int, int> goal) {
    int rows = grid.size(), cols = grid[0].size();
    auto cmp = [](Node* a, Node* b) { return *a > *b; };
    priority_queue<Node*, vector<Node*>, decltype(cmp)> open(cmp);
    vector<vector<bool>> closed(rows, vector<bool>(cols, false));

    Node* startNode = new Node(start.first, start.second);
    startNode->h = heuristic(start.first, start.second, goal.first, goal.second);
    open.push(startNode);

    while (!open.empty()) {
        Node* current = open.top();
        open.pop();

        int x = current->x, y = current->y;
        if (closed[x][y]) continue;
        closed[x][y] = true;

        if (x == goal.first && y == goal.second) {
            vector<pair<int, int>> path;
            while (current) {
                path.push_back({current->x, current->y});
                current = current->parent;
            }
            reverse(path.begin(), path.end());
            return path;
        }

        for (auto [dx, dy] : vector<pair<int, int>>{{0,1},{1,0},{0,-1},{-1,0}}) {
            int nx = x + dx, ny = y + dy;
            if (isValid(nx, ny, grid) && !closed[nx][ny]) {
                int g = current->g + 1;
                int h = heuristic(nx, ny, goal.first, goal.second);
                Node* neighbor = new Node(nx, ny, g, h, current);
                open.push(neighbor);
            }
        }
    }

    return {}; // No path found
}

int main() {
    vector<vector<int>> grid = {
        {0, 0, 0, 0},
        {1, 1, 0, 1},
        {0, 0, 0, 0},
        {0, 1, 1, 0}
    };

    pair<int, int> start = {0, 0};
    pair<int, int> goal = {3, 3};

    vector<pair<int, int>> path = astar(grid, start, goal);

    if (path.empty()) {
        cout << "No path found\n";
    } else {
        for (auto [x, y] : path)
            cout << "(" << x << "," << y << ") ";
        cout << "\n";
    }

    return 0;
}
